client/file_store_test.go

```diff
@@ -0,0 +1,103 @@
+package client
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+)
+
+func TestBadDirs(t *testing.T) {
+	// Note that since we create the temp directory, all the paths defined here
+	// in the tests are relative to that, as in, we just tack the md/td after
+	// it.
+	// Create the file store with the repository and repository/targets
+	// directories.
+	dir := t.TempDir()
+	os.Mkdir(filepath.Join(dir, "repository"), os.ModePerm)
+	os.Mkdir(filepath.Join(dir, "repository", "targets"), os.ModePerm)
+	os.Create(filepath.Join(dir, "repository-that-isfile"))
+	os.Create(filepath.Join(dir, "targets-that-isfile"))
+
+	tests := []struct {
+		name    string
+		md      string
+		td      string
+		wantErr string
+	}{{
+		name:    "empty, error",
+		wantErr: "metadataDir can't be empty",
+	}, {
+		name:    "invalid metadatadir, missing",
+		md:      "notthere",
+		wantErr: "notthere",
+	}, {
+		name:    "invalid targetsdir, missing",
+		md:      "repository",
+		td:      "targetsnotthere",
+		wantErr: "targetsnotthere",
+	}, {
+		name:    "invalid meta dir, is file",
+		md:      "repository-that-isfile",
+		wantErr: "not a directory",
+	}, {
+		name:    "invalid targets dir, is file",
+		md:      "targets-that-isfile",
+		wantErr: "not a directory",
+	}, {
+		name: "works, only specify meta",
+		md:   "repository",
+	}, {
+		name: "works, explicit targets",
+		md:   "repository",
+		td:   "repository/targets",
+	}}
+	for _, tc := range tests {
+		absoluteMeta := filepath.Join(dir, tc.md)
+		if tc.md == "" {
+			// pass in empty explicitly
+			absoluteMeta = tc.md
+		}
+		absoluteTargets := filepath.Join(dir, tc.td)
+		if tc.td == "" {
+			// pass in empty explicitly
+			absoluteTargets = tc.td
+		}
+		_, err := NewFileRemoteStore(absoluteMeta, absoluteTargets)
+		if tc.wantErr != "" && err == nil {
+			t.Errorf("%q wanted error %s, got none", tc.name, tc.wantErr)
+		} else if tc.wantErr == "" && err != nil {
+			t.Errorf("%q did not want error, got: %v", tc.name, err)
+		} else if err != nil && !strings.Contains(err.Error(), tc.wantErr) {
+			t.Errorf("%q wanted error %s but got: %s", tc.name, tc.wantErr, err)
+		}
+	}
+}
+
+// Test helper methods
+func (f *FileRemoteStore) addMeta(name string, data []byte) error {
+	return os.WriteFile(filepath.Join(f.metadataDir, name), data, os.ModePerm)
+}
+
+func (f *FileRemoteStore) addTarget(name string, data []byte) error {
+	return os.WriteFile(filepath.Join(f.targetsDir, name), data, os.ModePerm)
+}
+
+func (f *FileRemoteStore) deleteMeta(name string) error {
+	return os.Remove(filepath.Join(f.metadataDir, name))
+}
+
+func (f *FileRemoteStore) deleteTarget(name string) error {
+	return os.Remove(filepath.Join(f.targetsDir, name))
+}
+
+func newTestFileStore() (*FileRemoteStore, string, error) {
+	tmpDir := os.TempDir()
```

Comment:
i wish we could pass a testing tempdir here for both test cases! I don't really see how to do it via the Suite instances.

Reply:
I think that makes two of us :)